EssentialApp/EssentialAppTests/ImageCommentsUIIntegrationTests.swift

```diff
@@ -0,0 +1,259 @@
+		let comment2 = makeImageComment(id: UUID(), message: "message2", createdAt: (fixedDate.adding(days: -1), "1 day ago"), username: "username2")
+		let comment3 = makeImageComment(id: UUID(), message: "message3", createdAt: (fixedDate.adding(days: -2), "2 days ago"), username: "username3")
+		let comment4 = makeImageComment(id: UUID(), message: "message4", createdAt: (fixedDate.adding(days: -7), "1 week ago"), username: "username4")
+		
+		sut.loadViewIfNeeded()
+		assertThat(sut, isRendering: [])
+		
+		loader.completeImageCommentLoading(with: [comment0.model], at: 0)
+		assertThat(sut, isRendering: [comment0.viewModel])
+		
+		sut.simulateUserInitiatedReload()
+		loader.completeImageCommentLoading(with: [comment0.model, comment1.model, comment2.model, comment3.model, comment4.model], at: 1)
+		assertThat(sut, isRendering: [comment0.viewModel, comment1.viewModel, comment2.viewModel, comment3.viewModel, comment4.viewModel])
+	}
+	
+	func test_loadImageCommentCompletion_rendersSuccessfullyLoadedEmptyImageCommentsAfterNonEmptyImageComments() {
+		let (sut, loader) = makeSUT()
+		
+		let fixedDate = Date()
+		let comment0 = makeImageComment(id: UUID(), message: "message0", createdAt: (fixedDate.adding(seconds: -30), "30 seconds ago"), username: "username0")
+		let comment1 = makeImageComment(id: UUID(), message: "message1", createdAt: (fixedDate.adding(seconds: -30 * 60), "30 minutes ago"), username: "username1")
+		
+		sut.loadViewIfNeeded()
+		assertThat(sut, isRendering: [])
+		
+		loader.completeImageCommentLoading(with: [comment0.model, comment1.model], at: 0)
+		assertThat(sut, isRendering: [comment0.viewModel, comment1.viewModel])
+		
+		sut.simulateUserInitiatedReload()
+		loader.completeImageCommentLoading(with: [], at: 1)
+		assertThat(sut, isRendering: [])
+	}
+	
+	func test_loadImageCommentCompletion_doesNotAlterCurrentRenderingStateOnError() {
+		let (sut, loader) = makeSUT()
+		
+		let fixedDate = Date()
+		let comment0 = makeImageComment(id: UUID(), message: "message0", createdAt: (fixedDate.adding(seconds: -30), "30 seconds ago"), username: "username0")
+		let comment1 = makeImageComment(id: UUID(), message: "message1", createdAt: (fixedDate.adding(seconds: -30 * 60), "30 minutes ago"), username: "username1")
+		
+		sut.loadViewIfNeeded()
+		loader.completeImageCommentLoading(with: [comment0.model, comment1.model], at: 0)
+		assertThat(sut, isRendering: [comment0.viewModel, comment1.viewModel])
+		
+		sut.simulateUserInitiatedReload()
+		loader.completeImageCommentLoadingWithError(at: 1)
+		assertThat(sut, isRendering: [comment0.viewModel, comment1.viewModel])
+	}
+	
+	func test_loadImageCommentCompletion_rendersErrorMessageOnErrorUntilNextReload() {
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		loader.completeImageCommentLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, localized("COMMENT_VIEW_ERROR_MESSAGE"))
```

Comment:
Missing precondition:

```suggestion
		sut.loadViewIfNeeded()
		XCTAssertEqual(sut.errorMessage, nil)
		
		loader.completeImageCommentLoadingWithError(at: 0)
		XCTAssertEqual(sut.errorMessage, localized("COMMENT_VIEW_ERROR_MESSAGE"))
```